### PR TITLE
Add restrictions on names of domains - automated tests

### PR DESCRIPTION
**Pull request by XingY**
#### Related Pull Requests
- https://github.com/LabKey/labkey-ui-components/pull/1662
- https://github.com/LabKey/labkey-ui-premium/pull/616
- https://github.com/LabKey/limsModules/pull/984
- https://github.com/LabKey/platform/pull/6140
- https://github.com/LabKey/testAutomation/pull/2189
- https://github.com/LabKey/commonAssays/pull/826

#### Changes
-

### CODE REVIEW

File: src/org/labkey/test/tests/GpatAssayTest.java [hunk 1]
```diff
@@ -333,15 +335,20 @@ public void testMultipleFileUploadSingleRowInAssayRun()
     @Test
     public void testUpdateAssayDesign() throws IOException, CommandException
     {
+        File trialData = TestFileUtils.getSampleData("GPAT/renameAssayTrial.xls");
+
+        String invalidAssayName = TestDataGenerator.randomInvalidDomainName(10);
+        ReactAssayDesignerPage assayDesignerPage = startCreateGpatAssay(trialData, invalidAssayName);
+        List<String> errors = assayDesignerPage.clickSaveExpectingErrors();
+        assayDesignerPage.clickCancel();
+        Assert.assertTrue("Error msg not as expected during assay creation", errors.contains("Invalid Assay Design name. Domain name must start with a letter or a number character."));
```
Comment (labkey-susanh) [inline src/org/labkey/test/tests/GpatAssayTest.java]: Constant for the error message?

File: src/org/labkey/test/util/TestDataGenerator.java [hunk 2]
```diff
@@ -328,6 +330,31 @@ public static String randomString(int size, @Nullable String exclusion, @Nullabl
...
     }
 
+    public static String randomDomainName()
+    {
+        return randomDomainName(10);
+    }
+
+    public static String randomInvalidDomainName(int size)
+    {
+        return randomString(size, null, ILLEGAL_DOMAIN_NAME_CHARSET);
```
Comment (labkey-susanh) [inline src/org/labkey/test/util/TestDataGenerator.java:340]: Might be better (slightly more realistic) if this consisted of some number of illegal characters along with some legal characters. But perhaps the usages of this method already accomplish that.

File: src/org/labkey/test/util/TestDataGenerator.java [hunk 3]
```diff
@@ -328,6 +330,31 @@ public static String randomString(int size, @Nullable String exclusion, @Nullabl
...
+
+    public static String randomDomainName(int size)
+    {
+        String domainName = "";
+        do
+        {
+            String prefix = randomString(1, null, ALPHANUMERIC_STRING); // domain needs to start with alphanumeric char
+            final String charset = ALPHANUMERIC_STRING + DOMAIN_SPECIAL_STRING;
+            domainName = prefix + randomString(size - 1, null, charset);
+            domainName = domainName.trim();
```
Comment (labkey-susanh) [inline src/org/labkey/test/util/TestDataGenerator.java:351]: How would the name have any spaces to be trimmed?

File: src/org/labkey/test/util/TestDataGenerator.java [hunk 4]
```diff
@@ -328,6 +330,31 @@ public static String randomString(int size, @Nullable String exclusion, @Nullabl
...
+        }
+        while (domainName.length() < size || Pattern.matches("(.*\\s--[^ ].*)|(.*\\s-[^- ].*)", domainName)); // domain name must not contain space followed by dash. (command like)
```
Comment (labkey-susanh) [inline src/org/labkey/test/util/TestDataGenerator.java]: Would be good to reference the issue number here.

**Review by labkey-susanh — APPROVED** [review]

File: src/org/labkey/test/util/TestDataGenerator.java (hunk 3 above)
Comment (XingY) [reply src/org/labkey/test/util/TestDataGenerator.java:351]: They might have trailing space due to DOMAIN_SPECIAL_STRING contains whitespace

**Pull request merged**